hist/histv7/inc/ROOT/RBinIndexRange.hxx

```diff
@@ -0,0 +1,154 @@
+
+/**
+A range of bin indices \f$[fBegin, fEnd)\f$.
+
+The interface allows convenient iteration over RBinIndex. If included, RBinIndex::Underflow() is encountered before the
+normal bins and RBinIndex::Overflow() is the last value.
+
+~~~ {.cxx}
+ROOT::Experimental::RRegularAxis axis(10, 0, 1);
+for (auto index : axis.GetNormalRange(2, 5)) {
+   // Will iterate over [2, 3, 4]
+}
+for (auto index : axis.GetFullRange()) {
+   // Will iterate over all bins, starting with the underflow and ending with the overflow bin
+}
+~~~
+
+\warning This is part of the ROOT 7 prototype! It will change without notice. It might trigger earthquakes. Feedback is
+welcome!
+*/
+class RBinIndexRange final {
+   friend RBinIndexRange Internal::CreateBinIndexRange(RBinIndex, RBinIndex, std::size_t);
+
+   /// The begin of the range (inclusive)
+   RBinIndex fBegin;
+   /// The end of the range (exclusive)
+   RBinIndex fEnd;
+   /// The number of normal bins, after which iteration advances to RBinIndex::Overflow()
+   std::size_t fNNormalBins = 0;
+
+public:
+   /// Construct an invalid bin index range.
+   RBinIndexRange() = default;
+
+   RBinIndex GetBegin() const { return fBegin; }
+   RBinIndex GetEnd() const { return fEnd; }
+   // fNNormalBins is not exposed because it might be confusing for partial ranges.
+
+   friend bool operator==(const RBinIndexRange &lhs, const RBinIndexRange &rhs)
+   {
+      return lhs.fBegin == rhs.fBegin && lhs.fEnd == rhs.fEnd && lhs.fNNormalBins == rhs.fNNormalBins;
+   }
+
+   friend bool operator!=(const RBinIndexRange &lhs, const RBinIndexRange &rhs) { return !(lhs == rhs); }
+
+   /// Iterator over RBinIndex.
+   class Iterator final {
+      /// The current bin index
+      RBinIndex fIndex;
+      /// The number of normal bins, after which iteration advances to RBinIndex::Overflow()
+      std::size_t fNNormalBins = 0;
+
+   public:
+      using difference_type = std::size_t;
+      using value_type = RBinIndex;
+      using pointer = RBinIndex *;
+      using reference = RBinIndex &;
+      using iterator_category = std::forward_iterator_tag;
+
+      Iterator() = default;
+      Iterator(RBinIndex index, std::size_t nNormalBins) : fIndex(index), fNNormalBins(nNormalBins) {}
+
+      Iterator &operator++()
+      {
+         if (fIndex.IsUnderflow()) {
+            fIndex = 0;
+         } else if (fIndex.IsOverflow()) {
+            fIndex = RBinIndex();
+         } else if (fIndex.IsInvalid()) {
+            // This should never happen! In the worst case, when built with NDEBUG, the iterator stays at Invalid.
+            assert(0);
+         } else {
+            fIndex++;
+            if (fIndex.GetIndex() == fNNormalBins) {
+               fIndex = RBinIndex::Overflow();
+            }
+         }
+         return *this;
+      }
+      Iterator operator++(int)
+      {
+         Iterator old = *this;
+         operator++();
+         return old;
+      }
+
+      RBinIndex operator*() const { return fIndex; }
```

Comment:
I assume it's not a reference expecting it to be used mostly in C++20 and later?

Reply:
After quite some research, I changed the `iterator_category` to `std::input_iterator_tag` because it turns out that forward iterators have a requirement that we cannot fulfill:
https://eel.is/c++draft/forward.iterators#6 24.3.5.5.6
> If a and b are both dereferenceable, then a == b if and only if *a and *b are bound to the same object.

We cannot make this happen for member objects of the iterator.